docs/docs/proposals/003-grafanaserviceaccount-crd.md

```diff
@@ -0,0 +1,84 @@
+
+## Info
+
+Status: Suggested
+
+## Motivation
+
+Today Grafana Service Accounts has to be created after deploy when the grafana is running using the HTTP-API or the GUI. I instead suggest to have a Grafana Service Account CRD so that the Service Accounts could be predefined and created by the operator at deploy and the tokens will be created as k8s-secrets that then can be read by applications when needed.
+
+## Verification
+
+- Create integration tests for the operator creating grafana service accounts from a bare minimum yaml
+- Create integration tests for the operator creating grafana service accounts from a fully specified yaml
+- Create integration tests to check that it can rotate/invalidate tokens with TTL set (and passed).
+
+## Current solution
+
+Currently you are only able to create these grafana service accounts using the grafana GUI or by using the HTTP-API after the grafana has already been deployed and is running. And its removed when the grafana pod is restarted/redeployed without persistent storage. Meaning a new service account has to be manually created and its new token has to be updated where its being used.
+
+## Proposal
+
+My proposal is to handle grafana service account as a separate CRD that can be specified by the user even before setup and that can be included in a CICD pipeline. It will enable so that the operator can create predefined service accounts on deploy and store the token in a k8s-secret readable by other applications without any manual steps.
+
+### Defining what Grafana Service Account belongs to what grafana.
+I suggest that its done in the same way as dashboards and datasources are using instanceSelector to find the proper grafana, the same thing could be done with the grafana service accounts using the same labels etc.
+
+### Defining Grafana Service Account to Grafana operator.
+
+Today the Grafana Service Account is only held in memory if not using persistent, so when a pod is restarted or redeployed the grafana service account is removed. The grafana service accounts are also only possible to create using the Grafana GUI or with the HTTP-API and cannot be pre-defined before deploy or kept as IAC. But with the token kept as a k8s-secret in the same namespace as the GSA CR and the grafana instance it would be possible for other applications to use that token even when its rotated or recreated by the operator. RBAC rules are often more restrictive in the view scope so having the k8s-secret with the token in the same namespace as the grafana instance and other running applications would be beneficial.
+
+> Proposed CRD for GrafanaServiceAccounts
+
+```.yaml
+apiVersion: grafana.integreatly.org/v1beta1
+kind: GrafanaServiceAccount
+metadata:
+  labels:
+    app: grafana
+  name: grafana-sa
+  namespace: grafana-namespace
+spec:
+  create:
+    generateTokenSecret: [true/false] #Will create the k8s-secret with a default name if true. Defaults to true.
+  serviceaccount:
+    id: grafana-sa
```

Comment:
I know that we let our users set `id` in most places in the operator. Is it a good idea in the CRD though?
What would happen if two CRDs have the same ID? The tokens would be recreated over and over again, and it's a relatively big risk of creating an incident for the end users.

We could solve this by having admission controllers in the operator, the CEL language support isn't enough for separate CRDs. But adding this kind of functionality would make the operator much harder to get started with, since you need a ssl cert.

But we can just remove this issue by not letting the users setting this id and instead use the CRD name.

Reply:
I don't think it's super important to be able to set the id if that would cause issues. But it might be worth having the possibility to do it with a disclaimer of some sorts that this could happen if two of them would have the same ID? I feel that it isnt that different from working with for example flux where this happens if two resources have the same name etc. 

But that said, Im only basing it on how we use it in our organization, and I dont have a strong need for being able to set ID so I dont see an issue with excluding it.